### PR TITLE
Basic Json Support [NOT FINAL FEATURES]

### PR DESCRIPTION
**Pull request by mxwli**
# Description

Utility functions implemented: `to_json, json_merge_patch, json_extract, json_array_length, json_contains, json_keys, json_structure, json_valid, json`

Basic `LOAD FROM` implemented; scanning options not supported, scanning from multiple files not supported, `COPY FROM` not supported

Pull requests remaining:
1. Scanning Options
2. Json Writing

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: dataset/tinysnb/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+This dataset doesn't work as an actual database due to some limitations in JSON as well as Kuzu's casting system.
```
Comment (mxwli) [inline dataset/tinysnb/README.md:1]: I intend on making this database work with the next PR

File: src/main/connection.cpp [hunk 2]
```diff
@@ -19,6 +20,14 @@ Connection::Connection(Database* database) {
     KU_ASSERT(database != nullptr);
     this->database = database;
     clientContext = std::make_unique<ClientContext>(database);
+    if (!database->getConfig().readOnly) {
```
Comment (mxwli) [inline src/main/connection.cpp:23]: does this belong here? It seems to be causing ApiTest.ParallelConnect to hang.

File: dataset/tinysnb/README.md (hunk 1 above)
Comment (acquamarin) [inline dataset/tinysnb/README.md:1]: can you remove the readme file for now

File: third_party/CMakeLists.txt [hunk 3]
```diff
@@ -18,4 +18,5 @@ endif()
 add_subdirectory(re2)
 add_subdirectory(serd)
 add_subdirectory(utf8proc)
+add_subdirectory(yyjson)
```
Comment (acquamarin) [inline third_party/CMakeLists.txt:21]: yyjson is only used by the json extension, so i suggest moving json library under extension/json/third-party/

File: extension/json/src/json_extension.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+namespace kuzu {
+namespace json_extension {
+
+#define ADD_FUNC(FUNC_STRUCT)                                                                      \
+    extension::ExtensionUtils::registerFunctionSet(db, std::string(FUNC_STRUCT::name),             \
+        FUNC_STRUCT::getFunctionSet())
+
+void JsonExtension::load(main::ClientContext* context) {
+    auto& db = *context->getDatabase();
+    ADD_FUNC(ToJsonFunction);
```
Comment (acquamarin) [inline extension/json/src/json_extension.cpp:17]: Should register type there

File: third_party/CMakeLists.txt (hunk 3 above)
Comment (mxwli) [reply third_party/CMakeLists.txt]: This may be a problem; extension/json/third-party gets picked up by clang-tidy, and yyjson.c does not conform to clang-tidy requirements at all.

File: dataset/tinysnbjson/copy.cypher [hunk 5]
```diff
@@ -0,0 +1,10 @@
+load from 'dataset/tinysnbjson/vPerson.json' create (:person {ID: ID, fName: fName, gender: gender, isStudent: isStudent, isWorker: isWorker, age: age, eyeSight: eyeSight, birthdate:cast(birthdate as date), registerTime:cast(registerTime as timestamp), lastJobDuration:cast(lastJobDuration as interval), workedHours: workedHours, usedNames: usedNames, courseScoresPerTerm: courseScoresPerTerm, grades: grades, height: height, u: cast(u as uuid)});
```
Comment (acquamarin) [inline dataset/tinysnbjson/copy.cypher:1]: Should we suppport such grammar as well?
COPY person FROM 'vPerson.json' (ARRAY true);

Comment (acquamarin) [inline dataset/tinysnbjson/copy.cypher:1]: add an underscore between tinysnb and json in the directory name

File: extension/json/src/include/json_utils.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+namespace json_extension {
+
+struct JsonMutWrapper;
+
+class JsonWrapper {
+    JsonWrapper() : ptr{nullptr} {}
+
+public:
+    explicit JsonWrapper(yyjson_doc* ptr) : ptr{ptr} {}
+    ~JsonWrapper() {
```
Comment (acquamarin) [inline extension/json/src/include/json_utils.h:19]: Move to cpp

File: extension/json/src/include/json_utils.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    JsonWrapper(JsonWrapper&& other) { std::swap(ptr, other.ptr); }
+
+    yyjson_doc* ptr;
+};
+
+class JsonMutWrapper {
+public:
+    JsonMutWrapper() : ptr{yyjson_mut_doc_new(nullptr)} {}
+    explicit JsonMutWrapper(yyjson_mut_doc* ptr) : ptr{ptr} {}
+    ~JsonMutWrapper() {
```
Comment (acquamarin) [inline extension/json/src/include/json_utils.h:35]: move to cpp

File: extension/json/src/json_extension.cpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+#include "common/types/types.h"
+#include "json_functions.h"
+#include "json_scan.h"
+#include "main/client_context.h"
+#include "main/database.h"
+
+namespace kuzu {
+namespace json_extension {
+
+#define ADD_FUNC(FUNC_STRUCT)                                                                      \
```
Comment (acquamarin) [inline extension/json/src/json_extension.cpp:13]: I would move this macro to extension_utils.h so every extension can utilize this function

File: extension/json/src/json_extension.cpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    ADD_FUNC(JsonMergeFunction);
+    ADD_FUNC(JsonExtractFunction);
+    ADD_FUNC(JsonArrayLengthFunction);
+    ADD_FUNC(JsonContainsFunction);
+    ADD_FUNC(JsonKeysFunction);
+    ADD_FUNC(JsonStructureFunction);
+    ADD_FUNC(JsonValidFunction);
+    ADD_FUNC(MinifyJsonFunction);
+    extension::ExtensionUtils::registerTableFunction(db, JsonScan::getFunction());
+    if (!db.getCatalog()->containsType(&transaction::DUMMY_READ_TRANSACTION, "json")) {
```
Comment (acquamarin) [inline extension/json/src/json_extension.cpp:29]: I think createType should be able to handle duplicate json type, so you probably don't need the  check:
```
!db.getCatalog()->containsType(&transaction::DUMMY_READ_TRANSACTION, "json")
```

File: extension/json/src/json_functions.cpp [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+        result.setNull(resultPos, parameters[0]->isNull(inputPos));
+        if (!parameters[0]->isNull(inputPos)) {
+            StringVector::addString(&result, resultPos,
+                jsonToString(jsonify(*parameters[0], inputPos)));
+        }
+    }
+}
+
+function_set ToJsonFunction::getFunctionSet() {
+    function_set ret;
```
Comment (acquamarin) [inline extension/json/src/json_functions.cpp:30]: rename to result

File: extension/json/src/json_scan.cpp [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        }
+    } else {
+        rows.push_back(yyjson_doc_get_root(jsonBindData->json->ptr));
+    }
+    return std::make_unique<JsonScanSharedState>(std::move(rows));
+}
+
+static std::unique_ptr<TableFuncLocalState> initLocalState(TableFunctionInitInput&,
+    TableFuncSharedState* shared, storage::MemoryManager* /*mm*/) {
+    auto jsonShared = ku_dynamic_cast<TableFuncSharedState*, JsonScanSharedState*>(shared);
```
Comment (acquamarin) [inline extension/json/src/json_scan.cpp:116]: use shared->constCast<>() instead of ku_dynamic_cast

File: extension/json/src/json_scan.cpp [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    std::vector<yyjson_val*>::iterator begin, end;
+    uint32_t chunkSize;
+
+    JsonScanLocalState(std::vector<yyjson_val*>::iterator begin,
+        std::vector<yyjson_val*>::iterator end, uint32_t chunkSize)
+        : begin{begin}, end{end}, chunkSize{chunkSize} {}
+};
+
+struct JsonScanSharedState : public TableFuncSharedState {
+
```
Comment (acquamarin) [inline extension/json/src/json_scan.cpp:44]: remove space

File: src/binder/binder.cpp [hunk 13]
```diff
@@ -245,6 +245,10 @@ function::TableFunction Binder::getScanFunction(FileType fileType, const ReaderC
             csvConfig.parallel ? ParallelCSVScan::name : SerialCSVScan::name, inputTypes,
             functions);
     } break;
+    case FileType::JSON: {
```
Comment (acquamarin) [inline src/binder/binder.cpp:248]: This doesn't look right to me, json is an extension type and should not be known by the core(inlcuding binder).

File: src/binder/bind/read/bind_load_from.cpp [hunk 14]
```diff
@@ -83,6 +83,7 @@ std::unique_ptr<BoundReadingClause> Binder::bindLoadFrom(const ReadingClause& re
         case common::FileType::CSV:
         case common::FileType::PARQUET:
         case common::FileType::NPY:
+        case common::FileType::JSON:
```
Comment (acquamarin) [inline src/binder/bind/read/bind_load_from.cpp:86]: ditto, we should not explicitly handle json type in the core

File: src/include/common/constants.h [hunk 15]
```diff
@@ -200,5 +200,7 @@ static constexpr char LOCAL_DB_NAME[] = "local(kuzu)";
 
 constexpr auto DECIMAL_PRECISION_LIMIT = 38;
 
+static constexpr char SCAN_JSON_FUNC_NAME[] = "READ_JSON";
```
Comment (acquamarin) [inline src/include/common/constants.h:203]: Why do we put an extension function name in the core?

File: src/include/common/copier_config/reader_config.h [hunk 16]
```diff
@@ -17,6 +17,7 @@ enum class FileType : uint8_t {
     TURTLE = 4,   // Terse triples http://www.w3.org/TR/turtle
     NQUADS = 5,   // Line-based quads http://www.w3.org/TR/n-quads/
     NTRIPLES = 6, // Line-based triples http://www.w3.org/TR/n-triples/
+    JSON = 7,
```
Comment (acquamarin) [inline src/include/common/copier_config/reader_config.h:20]: remove, json is an extension function

File: src/include/main/client_context.h [hunk 17]
```diff
@@ -123,6 +123,10 @@ class KUZU_API ClientContext {
     void setDefaultDatabase(AttachedKuzuDatabase* defaultDatabase_);
     bool hasDefaultDatabase();
 
+    void runFuncInTransaction(const std::function<void(void)>& fun);
```
Comment (acquamarin) [inline src/include/main/client_context.h:126]: why did you move to public?

**Review by acquamarin — COMMENTED** [review]
Let's discuss how to avoid declaring the json type in the core

File: dataset/tinysnbjson/copy.cypher (hunk 5 above)
Comment (mxwli) [reply dataset/tinysnbjson/copy.cypher:1]: Yes, that's the plan, but scanning to a table isn't intended to be in this PR

File: src/binder/binder.cpp (hunk 13 above)
Comment (mxwli) [reply src/binder/binder.cpp:248]: Then how should the core invoke the JSON scan function?

File: src/include/main/client_context.h (hunk 17 above)
Comment (mxwli) [reply src/include/main/client_context.h:126]: This was originally used to register the json type, but now it's not necessary

File: extension/json/src/json_extension.cpp (hunk 9 above)
Comment (mxwli) [reply extension/json/src/json_extension.cpp:29]: catalog.createType does an assertion instead of handling it
```
    KU_ASSERT(!types->containsEntry(transaction, name));
```

**Review by andyfengHKU — APPROVED** [review]
